### PR TITLE
All of queries should return correct result even if including large number

### PR DESCRIPTION
**Pull request by kamipo**
Currently several queries cannot return correct result due to incorrect
`RangeError` handling.

First example:

```ruby
assert_equal true, Topic.where(id: [1, 9223372036854775808]).exists?
assert_equal true, Topic.where.not(id: 9223372036854775808).exists?
```

The first example is obviously to be true, but currently it returns
false.

Second example:

```ruby
assert_equal topics(:first), Topic.where(id: 1..9223372036854775808).find(1)
```

The second example also should return the object, but currently it
raises `RecordNotFound`.

It can be seen from the examples, the queries including large number
assuming empty result is not always correct.

Therefore, This change handles `RangeError` to generate executable SQL
instead of raising `RangeError` to users to always return correct
result. By this change, it is no longer raised `RangeError` to users.

### CODE REVIEW

**Comment by matthewd** [comment]
Ugh... I'm sure I was worried about these sorts of cases when we added that rescue :confused:

I don't think replacing binds with inline values is the solution, though. It provides a trivial means of blowing out the prepared statement cache, and even in the best case gives a surprise inefficient query.. what was an index lookup is now a full table scan.

If we can't pick out just the known cases, like `find(hugenum)`, I think we have to just let the exception raise.

**Comment by kamipo** [comment]
I changed the implementation to build `"1=0"` predicate rather than replacing binds with inline values. I think it is no longer given inefficient query.

**Comment by rafaelfranca** [comment]
I'm thing this new implementation is good. @matthewd do you still have concerns?

File: activerecord/lib/active_record/relation/predicate_builder.rb
Comment (sgrif) [inline activerecord/lib/active_record/relation/predicate_builder.rb]: Why do we have the random `nil?` check?

**Review by sgrif — APPROVED** [review]
`build_query_attribute` appears to be doing something really implicit. I'd like to see it be made much more explicit, but this seems fine to me overall. The additional test cases make sense.

I would like to see the intent of `build_query_attribute` made more clear before this is merged though

File: activerecord/lib/active_record/relation/predicate_builder.rb
Comment (kamipo) [reply activerecord/lib/active_record/relation/predicate_builder.rb]: This intend to call `QueryAttribute#value_for_database` in `QueryAttribute#nil?` in `BindParam` to check `Type::Integer#ensure_in_range` before building arel predicate.

https://github.com/rails/rails/blob/51c5bcbb39777b3861a44042a598bfb07d4e1fc1/activerecord/lib/active_record/relation/query_attribute.rb#L20-L23

Comment (sgrif) [reply activerecord/lib/active_record/relation/predicate_builder.rb]: See my second comment. I'd like to see this made more explicit in the code. This is too spooky-action-at-a-distance for my taste.

Comment (sgrif) [reply activerecord/lib/active_record/relation/predicate_builder.rb]: Especially since `nil?` there is basically just a hack to satisfy Arel that I'd like to remove at some point.

Comment (kamipo) [reply activerecord/lib/active_record/relation/predicate_builder.rb]: hm... how about here?

```diff
diff --git a/activerecord/lib/active_record/relation/predicate_builder.rb b/activerecord/lib/active_record/relation/predicate_builder.rb
index c69a238eff..56e632f0d7 100644
--- a/activerecord/lib/active_record/relation/predicate_builder.rb
+++ b/activerecord/lib/active_record/relation/predicate_builder.rb
@@ -58,7 +58,13 @@ def build_bind_attribute(column_name, value)
 
     def build_query_attribute(column_name, value)
       bind = build_bind_attribute(column_name, value)
-      bind.nil?
+      attr = bind.value
+
+      # Checking whether an integer value is within the range
+      if attr.type.type == :integer && attr.value_before_type_cast
+        attr.value_for_database
+      end
+
       bind
     rescue ::RangeError
     end
```

File: activerecord/test/cases/finder_test.rb
Comment (matthewd) [inline activerecord/test/cases/finder_test.rb]: Some more: 
```ruby
assert_equal false, Topic.where(id: 9223372036854775808..9223372036854775809).exists?
assert_equal true, Topic.where(id: -9223372036854775809..9223372036854775808).exists?
```

(I'm pretty sure these would both currently be `true`?)

File: activerecord/lib/active_record/relation/predicate_builder/basic_object_handler.rb
Comment (matthewd) [inline activerecord/lib/active_record/relation/predicate_builder/basic_object_handler.rb]: I think `attribute.in([])` would be more consistent with the range handler, as a more abstract false.

File: activerecord/lib/active_record/relation/predicate_builder/relation_handler.rb
Comment (matthewd) [inline activerecord/lib/active_record/relation/predicate_builder/relation_handler.rb]: This doesn't seem related?

File: activerecord/lib/active_record/relation/predicate_builder.rb
Comment (matthewd) [reply activerecord/lib/active_record/relation/predicate_builder.rb]: That's slightly more obvious about what it's doing, but still feels both over-specialised and too vague at the same time. My first thought was maybe we want a `bind.value.impossible?` or similar, but that isn't actually sufficient for our needs: see the new tests I proposed.

I'm not excited about this method name, either... particularly the fact we're losing "bind", which seems important to understand while reading callers. But that's better revisited after the changes for the new tests too.

**Review by matthewd — COMMENTED** [review]
This still opens the possibility of 2**N query cache entries for a single query with N integer parameters, but that's far more bounded than infinity. And it avoids wasting database effort on the known-false condition, so that's good.

I agree the current rescues are catching too much, so something must change. If this can fix it while avoiding even more exceptions, so much the better.

File: activerecord/lib/active_record/relation/predicate_builder/relation_handler.rb
Comment (kamipo) [reply activerecord/lib/active_record/relation/predicate_builder/relation_handler.rb]: Right, unrelated. I removed the change.

File: activerecord/lib/active_record/relation/predicate_builder/basic_object_handler.rb
Comment (kamipo) [reply activerecord/lib/active_record/relation/predicate_builder/basic_object_handler.rb]: Yeah, changed to `attribute.in([])`.

File: activerecord/test/cases/finder_test.rb
Comment (kamipo) [inline activerecord/test/cases/finder_test.rb]: > (I'm pretty sure these would both currently be `true`?)

Good point. I've fixed and added the tests.

File: activerecord/lib/active_record/relation/predicate_builder.rb
Comment (kamipo) [inline activerecord/lib/active_record/relation/predicate_builder.rb]: I've changed it from `bind.nil?` for now.

File: activerecord/lib/active_record/relation/predicate_builder/range_handler.rb
Comment (matthewd) [inline activerecord/lib/active_record/relation/predicate_builder/range_handler.rb]: I think converting all out-of-range values to infinity is still problematic. Try `9223372036854775808..10`.

I'm also worried that this class really shouldn't be doing any integer/integer-like operations. Just because something is infinite doesn't mean it's comparable with zero.

How about we push the other direction, and instead take *out* the concept of infinity? The question this code needs to ask is whether the value is above- or below range: Infinity is both of those (because we give it special treatment), but over-sized integer values are one or the other.

After thinking through a couple of possible APIs, the least bad one I can come up with is to add an `impossible_database_value?` method to AMo::Attribute, with documented return values of `false` (value is normal), `true` (value is impossible, no more information available), `:above`, and `:below` (value is greater/less than all possible values). Then values where `value_for_database` will raise RangeError should return one of the truthy values from that method.

@sgrif this is extending your nice simple low level interface in way I'm not particularly proud of; opinions / better ideas very welcome.

Comment (kamipo) [reply activerecord/lib/active_record/relation/predicate_builder/range_handler.rb]: > I think converting all out-of-range values to infinity is still problematic. Try `9223372036854775808..10`.

ah... I've addressed it in e80e677 anyway.

Originally this PR was written as not to add any APIs as much as possible (the converting to infinity, and `bind.nil?`).
But I agree that adding an API for this issue will make it easier to understand what problems are addressed.
So I'll try another direction to make more explicit in the code.

**Review by rafaelfranca — APPROVED** [review]
Sounds good to me, but let's wait @matthewd's approval

File: activerecord/test/cases/finder_test.rb
Comment (kamipo) [inline activerecord/test/cases/finder_test.rb]: Added test case for open-ended range.

File: activerecord/lib/active_record/relation/predicate_builder/range_handler.rb
Comment (kamipo) [inline activerecord/lib/active_record/relation/predicate_builder/range_handler.rb]: I've removed converting out-of-range values to infinity.
So this line will check out-of-range values only, doesn't affect open-ended ranges.

File: activerecord/lib/active_record/relation/predicate_builder/range_handler.rb
Comment (matthewd) [inline activerecord/lib/active_record/relation/predicate_builder/range_handler.rb]: This really feels like it's taking more effort to work around the problem than giving the attribute responsibility for ranges.

My examples are getting more obscure, and the gaps are getting smaller... but the fact there are *any* gaps highlights that it feels like a piecemeal solution, with lots of conditionals, instead of few wide-reaching rules.

```ruby
x = Author.create! name: "12 Monkeys"
assert_equal [x], Author.where(name: 10..2)
```

Maybe that's silly enough that we don't care.. but it is a change in behaviour. 

(Or maybe I'm misreading and the value's already been cast before this? But I think that only happens below.)

I'll defer to @sgrif, but I still feel like we want something like https://github.com/rails/rails/pull/30000#discussion_r162247130 instead of having this class expect more from all the value types it might encounter.

File: activerecord/lib/active_record/relation/predicate_builder/range_handler.rb
Comment (kamipo) [inline activerecord/lib/active_record/relation/predicate_builder/range_handler.rb]: Thanks for the feedback. I've added `boundable?` and `impossible?` helper methods for the range handler to make more explicit in the code.

File: activerecord/test/cases/finder_test.rb
Comment (kamipo) [inline activerecord/test/cases/finder_test.rb]: Added test case for numeric range for string column.

**Comment by kamipo** [comment]
In this PR, finder methods no longer raise `RangeError`.
So `StatementCache#execute` is the only place to raise the exception for finder queries.

#29988 closely related with this PR to avoid catching the exception in much places.
Thus I cherry-picked #29988 in this PR.

File: activerecord/lib/active_record/statement_cache.rb
Comment (matthewd) [inline activerecord/lib/active_record/statement_cache.rb]: This rescue makes me nervous, just because it *sounds* far-reaching.

I agree it's fine to put it here in practice, because of the very limited set of statements that are eligible for this cache, as you noted in the commit message. I think that's worth a comment here too, though: both to hopefully draw attention if the cache ever gets used for other queries, and just to reassure ~future readers~ me next time I read through this code 😅

File: activerecord/lib/active_record/statement_cache.rb
Comment (kamipo) [inline activerecord/lib/active_record/statement_cache.rb]: Added the comment for returning empty result.
@matthewd Is it enough for you?

File: activerecord/lib/active_record/relation/predicate_builder/range_handler.rb
Comment (kamipo) [inline activerecord/lib/active_record/relation/predicate_builder/range_handler.rb]: TBH I'd like to postpone implementing like `impossible_database_value?` API (https://github.com/rails/rails/pull/30000#discussion_r162247130) in this PR.

The API is only useful for building `>` or `<` with boundable queries (like #31219), so range predicate handler is the only place where it will be used for now.

But the range predicate handler need to treat infinity value as open-ended regardless of whether + or - value. Even if `impossible_database_value?` is implemented, it does not make the code easy.

@matthewd @sgrif What do you think about that?

```diff
diff --git a/activerecord/lib/active_record/relation/predicate_builder/range_handler.rb b/activerecord/lib/active_record/relation/predicate_builder/range_handler.rb
index 7ea9b46ab6..4497d9e296 100644
--- a/activerecord/lib/active_record/relation/predicate_builder/range_handler.rb
+++ b/activerecord/lib/active_record/relation/predicate_builder/range_handler.rb
@@ -17,7 +17,7 @@ def call(attribute, value)
         begin_bind = predicate_builder.build_bind_attribute(attribute.name, value.begin)
         end_bind = predicate_builder.build_bind_attribute(attribute.name, value.end)
 
-        if impossible?(begin_bind) && value.begin > 0 || impossible?(end_bind) && value.end < 0
+        if impossible?(begin_bind) && begin_bind.value.impossible_database_value? == :above || impossible?(end_bind) && end_bind.value.impossible_database_value? == :below
           attribute.in([])
         elsif infinity?(value.begin) || !begin_bind.value.boundable?
           if infinity?(value.end) || !end_bind.value.boundable?
diff --git a/activerecord/lib/active_record/relation/query_attribute.rb b/activerecord/lib/active_record/relation/query_attribute.rb
index 6b96cf6745..f63846815b 100644
--- a/activerecord/lib/active_record/relation/query_attribute.rb
+++ b/activerecord/lib/active_record/relation/query_attribute.rb
@@ -29,6 +29,16 @@ def boundable?
       rescue ::RangeError
         @_boundable = false
       end
+
+      def impossible_database_value?
+        if boundable?
+          false
+        elsif value_before_type_cast > 0
+          :above
+        else
+          :below
+        end
+      end
     end
   end
 end
```

**Comment by kamipo** [comment]
I'd be going to merge this PR since I think I've addressed all suggestions, and this PR already got two approvals and the intent of the code is clear enough.

Let me know if anyone has any suggestions, I think I could address that before the final 6.0 release.

cc @rafaelfranca @matthewd

**Comment by rafaelfranca** [comment]
:shipit:

**Pull request merged**
